src/dataflow/src/source/mod.rs

```diff
@@ -849,6 +853,221 @@ enum MessageProcessing {
     YieldedWithDelay,
 }
 
+type EventSender =
+    mpsc::Sender<Event<Option<Timestamp>, Result<(Row, Timestamp, Diff), SourceError>>>;
+
+/// An active transaction at a particular point in time. An instance of this struct is provided to
+/// a source when calling start_tx() on its timestamper. This has the effect of freezing the
+/// timestamper clock while the data for the transaction is sent.
+///
+/// The transaction is automatically committed on Drop at which point the timestamper will continue
+/// ticking its internal clock.
+pub struct SourceTransaction<'a> {
+    timestamp: RwLockReadGuard<'a, Timestamp>,
+    sender: &'a EventSender,
+}
+
+impl SourceTransaction<'_> {
+    /// Record an insertion of a row in the current transaction
+    pub async fn insert(&self, row: Row) -> anyhow::Result<()> {
+        let msg = Ok((row, *self.timestamp, 1));
+        self.sender
+            .send(Event::Message(*self.timestamp, msg))
+            .await
+            .or_else(|_| Err(anyhow!("channel closed")))
+    }
+
+    /// Record a deletion of a row in the current transaction
+    pub async fn delete(&self, row: Row) -> anyhow::Result<()> {
+        let msg = Ok((row, *self.timestamp, -1));
+        self.sender
+            .send(Event::Message(*self.timestamp, msg))
+            .await
+            .or_else(|_| Err(anyhow!("channel closed")))
+    }
+}
+
+/// A thread-safe transaction manager that is responsible for assigning timestamps to the
+/// transactions submitted to the system. Rows can be inserted individually using the
+/// [insert](Timestamper::insert) and [delete](Timestamper::delete) methods or as part of a bigger
+/// transaction.
+///
+/// When a transaction is started using [start_tx](Timestamper::start_tx) the internal clock will be
+/// frozen and any subsequent rows will be timestamped with the exact same timestamp. The
+/// transaction is committed automatically as soon as the transaction object gets dropped.
+pub struct Timestamper {
+    inner: Arc<RwLock<Timestamp>>,
+    sender: EventSender,
+    tick_duration: Duration,
+}
+
+impl Timestamper {
+    fn new(sender: EventSender, tick_duration: Duration) -> Self {
+        let now = UNIX_EPOCH
+            .elapsed()
+            .expect("system clock before 1970")
+            .as_millis()
+            .try_into()
+            .expect("materialize has existed for more than 500M years");
+        Self {
+            inner: Arc::new(RwLock::new(now)),
+            sender,
+            tick_duration,
+        }
+    }
+
+    /// Start a transaction at a particular point in time. The timestamper will freeze its internal
+    /// clock while a transaction is active.
+    pub async fn start_tx<'a>(&'a self) -> SourceTransaction<'a> {
```

Comment:
I just want to double check that I understand -- if a user has 3 transactions A, B, and C that all show up at the same second, but if the timestamp frequency is say 1 minute then all of those transactions would get bound to the same timestamp right? (as far as I can tell this doesn't mean that anything is super broken, except that if someone wanted to observe the effects of _just_ transaction A they couldn't -- is that right?)

Reply:
That's right, yeah. But this sounds almost like a feature? As in, consistency is maintained and the frequency interval parameter defines how fine grained your time axis is